include/sampleflow/producers/differential_evaluation_mh.h

```diff
@@ -0,0 +1,201 @@
+            const std::function<OutputType (const OutputType &, const OutputType &, const OutputType &)> &crossover,
+            const unsigned int crossover_gap,
+            const types::sample_index n_samples)
+    {
+      unsigned int n_chains = starting_points.size();
+      assert (n_chains >= 3);
+      // Make sure the flush_consumers() function is called at any point
+      // where we exit the current function.
+      Utilities::ScopeExit scope_exit ([this]()
+      {
+        this->flush_consumers();
+      });
+
+      std::mt19937 rng;
+      // Initialize distribution for comparing to acceptance ratio
+      std::uniform_real_distribution<> uniform_distribution(0,1);
+
+      std::valarray<OutputType> current_samples = starting_points;
+      // Include another array to store new values so that all crossovers
+      // can be performed with the previous set of samples
+      std::valarray<OutputType> next_samples = starting_points;
+
+      // Loop over the desired number of samples
+      for (types::sample_index i=0; i<n_samples/n_chains; ++i)
+        {
+          // Loop over the desired number of chains
+          for (unsigned int chain = 0; chain < n_chains; ++chain)
+            {
+              // Determine trial sample and likelihood ratio; either from
+              // crossover operation or regular perturbation
+              std::pair<OutputType, double> trial_sample_and_ratio;
+              // Perform crossover every crossover_gap iterations
+              if ((i % crossover_gap) == 0)
+                {
+                  // Select two chains to combine
+                  std::uniform_int_distribution<> a_dist(0, n_chains - 2);
+                  int a = a_dist(rng);
+                  if (a >= i)
+                    a += 1;
+                  OutputType trial_a = current_samples[a];
+                  std::uniform_int_distribution<> b_dist(0, n_chains - 3);
+                  int b = b_dist(rng);
+                  if (b >= std::max<int>(a, chain))
+                    b += 2;
+                  else if (b >= std::min<int>(a, i))
+                    b += 1;
+                  OutputType trial_b = current_samples[b];
+                  // Combine trial a and trial b
+                  OutputType crossover_result = crossover(current_samples[chain], trial_a, trial_b);
+                  trial_sample_and_ratio = perturb(crossover_result);
+                }
+              else
+                trial_sample_and_ratio = perturb(current_samples[chain]);
+
+              OutputType trial_sample = std::move(trial_sample_and_ratio.first);
+              const double proposal_distribution_ratio = trial_sample_and_ratio.second;
+              const double trial_log_likelihood = log_likelihood (trial_sample);
+              // Accept trial sample with probability equal to ratio of likelihoods;
+              // (always accept if > 1)
+              double current_log_likelihood = log_likelihood(current_samples[chain]);
```

Comment:
Instead of re-computing the log likelihood each iteration, store a vector of current log likelihoods and only re-compute when a new sample is accepted for each chain. Or, instead of a vector of current samples, store a vector of pairs (current sample, current log likelihood).